core/ir/synthetic/encode.c

```diff
@@ -0,0 +1,213 @@
+                    src_reg_used[reg_canonical] = true;
+                    if (reg_size_in_bytes > max_reg_size)
+                        max_reg_size = reg_size_in_bytes;
+                }
+            }
+        } else {
+            for (uint opnd_index = 0; opnd_index < num_regs_used_by_opnd; ++opnd_index) {
+                reg_id_t reg = opnd_get_reg_used(dst_opnd, opnd_index);
+                /* Map sub-registers to their containing register.
+                 */
+                reg_id_t reg_canonical = reg_to_pointer_sized(reg);
+                opnd_size_t reg_size = reg_get_size(reg_canonical);
+                uint reg_size_in_bytes = opnd_size_in_bytes(reg_size);
+                if (!dst_reg_used[reg_canonical]) {
+                    ++num_dsts;
+                    dst_reg_used[reg_canonical] = true;
+                    if (reg_size_in_bytes > max_reg_size)
+                        max_reg_size = reg_size_in_bytes;
+                }
+            }
+        }
+    }
+    encoding_header |= num_dsts;
+
+    /* Encode number of register source operands, adding on top of already existing ones.
+     */
+    uint original_num_srcs = (uint)instr_num_srcs(instr);
+    for (uint i = 0; i < original_num_srcs; ++i) {
+        opnd_t src_opnd = instr_get_src(instr, i);
+        uint num_regs_used_by_opnd = (uint)opnd_num_regs_used(src_opnd);
+        for (uint opnd_index = 0; opnd_index < num_regs_used_by_opnd; ++opnd_index) {
+            reg_id_t reg = opnd_get_reg_used(src_opnd, opnd_index);
+            /* Map sub-registers to their containing register.
+             */
+            reg_id_t reg_canonical = reg_to_pointer_sized(reg);
+            opnd_size_t reg_size = reg_get_size(reg_canonical);
+            uint reg_size_in_bytes = opnd_size_in_bytes(reg_size);
+            if (!src_reg_used[reg_canonical]) {
+                ++num_srcs;
+                src_reg_used[reg_canonical] = true;
+                if (reg_size_in_bytes > max_reg_size)
+                    max_reg_size = reg_size_in_bytes;
+            }
+        }
+    }
+    encoding_header |= (num_srcs << SRC_OPND_SHIFT);
+
+    /* Encode arithmetic flags.
+     * In AARCH64 instr_get_arith_flags() may need to do additional decoding of instr.
+     * If that happens, #dcontext_t ISA mode will be set to be the #intr_t ISA mode, which
+     * is DR_ISA_REGDEPS.
+     * This would trigger the decoding for DR_ISA_REGDEPS, which is not what we want.
+     * So, we temporary set instr ISA mode to be the dcontex_t ISA mode, which contains
+     * a real ISA mode (e.g., DR_ISA_ARM_A64).
+     */
+    dr_isa_mode_t dcontext_isa_mode = dr_get_isa_mode((void *)dcontext);
+    CLIENT_ASSERT(dcontext_isa_mode != DR_ISA_REGDEPS,
+                  "dcontext_t ISA mode cannot be DR_ISA_REGDEPS when encoding into a "
+                  "synthetic instruction");
+    instr_set_isa_mode(instr, dcontext_isa_mode);
+    uint eflags_instr = instr_get_arith_flags(instr, DR_QUERY_DEFAULT);
```

Comment:
This might trigger additional decoding, which uses dcontext_t ISA mode (retrieved from instr_t ISA mode).
This also means that in the (coming soon) encoding_filter we'll have to guard encoding to synthetic instructions with mutexes, since that happens in parallel (per-shard) and we're changing the "global" ISA of dcontext_t even in decoding.

Reply:
Wait a minute: I think there is something here we may not be able to support.  Let me make sure I understand how the user does this:

Scenario A which has a fully decoded Level 3 instr:

+ Call decode() with dcontext set to x86_64 say which sets the instr mode to x86_64 and fills in eflags info and sets INSTR_EFLAGS_VALID.
+ Call instr_set_isa_mode() to set to REGONLY. 
   + Aside: this should clear the raw bits since those are no longer valid: looks like it doesn't do that today -- should that be part of this PR which is adding this cross-arch scenario for the first time?)
+ Now call an instr_encode method.  This will not invoke any up-decoding.

Scenario B: similar, but start with a synthesized Level 4 instr from INSTR_CREATE_* or something: has no eflags info; has INSTR_EFLAGS_VALID unset.  A call to instr_get_eflags() will go into the needs-encoding branch and (recursively) call instr_encode on REGONLY won't it?  This is from me glancing at the source code.  This doesn't look good.

But if we don't support encoding to REGONLY without INSTR_EFLAGS_VALID: do we not support encoding Level 4 to REGONLY at all?  The user has to first encode to a real ISA?  There should be an API routine to combine the two calls like instr_encode_to_isa()??

Maybe we should discuss offline.

Reply:
Scenario A (which is what should happen in the encoding_filter):
I agree with what you said, this should not invoke any additional decoding, hence it should be fine.

Scenario B (which is what happens in suite/tests/api/ir_synthetic.c):
As you said, we have an instr_t coming from INSTR_CREATE_*() of some real ISA (let's say AARCH64, since that's where this issue happened... although, it might still be a problem for other arches as well).
INSTR_EFLAGS_VALID is unset.
We set the instr_t ISA mode to be DR_ISA_REGDEPS, so when we call instr_encode() on it we trigger a call to encode_to_synth(), where we take an instr_t from a real ISA and convert it to an encoded synthetic instruction as a sequence of bytes (aligned by 4).
In encode_to_synth() we call instr_get_arith_flags() on the instr_t we are converting (which again, is technically a real ISA instr_t, but has its ISA mode set to DR_ISA_REGDEPS, which is why we're having these issues).
encode_to_synth() calls instr_get_eflags(), which sets dcontext_t ISA mode to be the same as the given instr_t ISA mode (DR_ISA_REGDEPS in this case) and then calls decode_eflags_usage(), which (at least in AARCH64) calls decode_common() with a dcontext_t ISA mode now set to DR_ISA_REGDEPS.
decode_common() now triggers a call to decode_from_synth() which will operate on an encoded instruction that has the encoding of a real ISA (AARCH64), not our synthetic one... so things break.

> A call to instr_get_eflags() will go into the needs-encoding branch and (recursively) call instr_encode
You mean "will go into needs-decoding branch and (recursively) call decode*()"?
If so, yes, as explained above.
I don't think instr_get_eflags() encodes anything (might be wrong here though), but it can call decode_common() (if AARCH64) or read_instruction() (if x86).

About the other issues you pointed out:

> But if we don't support encoding to REGONLY without INSTR_EFLAGS_VALID
I think we do, now that we temporarily switch back the instr_t ISA mode to a real one.
If you mean: this is too ugly and fragile to be a solution, I agree.

> The user has to first encode to a real ISA?
You mean "decode to a real ISA"? Yes.
I think this is the main issue.
DR_ISA_REGDEPS works this way: encode (takes a real ISA instr_t, outputs a synthetic encoded instruction), decode (takes a synthetic encoded instruction, outputs a synthetic ISA instr_t).
While all the other ISA(s) work this way: encode (takes a real ISA instr_t, outputs the same real ISA encoded instruction) decode (takes a real ISA encoded instruction, outputs the same real ISA instr_t).
The asymmetry between the two is fundamentally the problem.

> There should be an API routine to combine the two calls like instr_encode_to_isa()??
Not super sure about what you mean here.
Do you mean we should have a "converter" from a real ISA instr_t to a DR_ISA_REGDEPS instr_t?
That is probably a very good idea now that I think about it, and it would remove the asymmetry between encoding/decoding of DR_ISA_REGDEPS vs. all the other real ISA(s).

Reply:
> encode_to_synth() calls instr_get_eflags(), which sets dcontext_t ISA mode to be the same as the given instr_t ISA mode (DR_ISA_REGDEPS in this case) and then calls decode_eflags_usage(),

Why doesn't it first call private_instr_encode(), as the instr needs encoding?  Won't that then recurse?

Reply:
See https://github.com/DynamoRIO/dynamorio/blob/master/core/ir/instr_shared.c#L1060

Reply:
Re: "will go into needs-decoding" and "You mean "decode"": No I mean encode as I said: Level 4 cannot be decoded as there are no raw bits to decode.  Only after encoding can you do decoding if you have Level 4.

Reply:
> Why doesn't it first call private_instr_encode(), as the instr needs encoding? Won't that then recurse?

It does, but private_instr_encode() calls instr_encode_check_reachability() and, if needed, instr_encode_ignore_reachability(), and both call instr_encode_arch(), hence they bypass encode_to_synth() (which is only reachable from instr_encode() and instr_encode_to_copy()).
So, private_instr_encode() does the "right thing" and doesn't trigger encode_to_synth().

Reply:
This does seem fragile and messy with these mixed types of encodings on the same call paths.  Switching to a completely separate ISA conversion step, and then having the ISA encode paths clearly separated, as discussed offline, seems the cleaner approach.

The REGDEPS ISA would then require INSTR_EFLAGS_VALID to be set I guess and users synthesizing at Level 4 in the REGDEPS ISA would have to manually set the flags.  REGDEPS would also fail to encode an instr_t with memops.

Reply:
The conversion step (now done by: instr_convert_to_isa_regdeps()) takes care of setting INSTR_EFLAGS_VALID for  the instruction converted to the synthetic ISA "DR_ISA_REGDEPS".
DR_ISA_REGDEPS encoding should not be triggered by instructions that belong to a real ISA (their ISA mode is != than DR_ISA_REGDEPS), which are the only ones that can have memops.
By "definition", DR_ISA_REGDEPS instructions (which are the only ones encode_to_synth() operates on) do not have memops.

Reply:
We now only set the arithmetic flags to be valid in the conversion step, and assert that they are in encode.